x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -192,16 +206,37 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
             // An error occurred before we got the shard list
             searchResponse = null;
         } else {
-            /*
-             * Build the response, reducing aggs if we haven't already and
-             * storing the result of the reduction, so we won't have to reduce
-             * the same aggregation results a second time if nothing has changed.
-             * This does cost memory because we have a reference to the finally
-             * reduced aggs sitting around which can't be GCed until we get an update.
-             */
-            InternalAggregations reducedAggs = reducedAggsSource.get();
-            reducedAggsSource = () -> reducedAggs;
-            searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            // partial results branch
+            SearchResponseMerger searchResponseMerger = task.getSearchResponseMerger();
```

Comment:
Looking at this, I am also wondering if instead of retrieving the merger from the task we should rather move the logic to the task and expose some `merge` method instead on the task itself. I don't have a strong opinion though. I am looking for ways to simplify this block and make it more readable, yet moving it does not necessarily help that :)

Reply:
I think putting it in SearchTask doesn't simplify the logic (as you say) and to me adds in unnecessary role/complexity to AsyncSearchTask, as we'd basically have to have the same merge logic but now in a less direct place?